File: Dec2020/3Dec/MinimumIndexSumOfTwoLists.cpp
/*
Problem name: 599. Minimum Index Sum of Two Lists
Problem link: https://leetcode.com/problems/minimum-index-sum-of-two-lists/
*/

class Solution {
public:
    vector<string> findRestaurant(vector<string>& list1, vector<string>& list2) {
        map<string, int> l1;
        int minv=INT_MAX,indexsum;
        vector<string> ans;
        
        for(int i=0;i<list1.size();++i)
            l1[list1[i]]=i;
        
        for(int i=0;i<list2.size();++i){
            if(l1.find(list2[i]) != l1.end()){
                
                indexsum = l1[list2[i]] + i;
                
                if(indexsum < minv)
                    ans.clear();
                
                if(indexsum <= minv){
                    minv = indexsum;
                    ans.push_back(list2[i]);
                }
            }
        }
        return ans;
    }
};

File: Dec2020/3Dec/MinimumIndexSumOfTwoLists.java
/*
Problem Name: 599. Minimum Index Sum of Two Lists
Problem Link: https://leetcode.com/problems/minimum-index-sum-of-two-lists/
*/
class Solution {
    public String[] findRestaurant(String[] list1, String[] list2) {
        HashMap<String,Integer> map1 = new HashMap<String,Integer>();
        HashMap<String,Integer> map2 = new HashMap<String,Integer>();
        for(int i=0;i<list1.length;i++)
            map1.put(list1[i],i);
        for(int i=0;i<list2.length;i++)
            map2.put(list2[i],i);
        HashMap<String,Integer> common = new HashMap<String,Integer>();
        int min = Integer.MAX_VALUE;
        for(String s:map1.keySet()){
            if(map2.containsKey(s)){
                common.put(s, map1.get(s) + map2.get(s));
                min = Math.min(min, common.get(s));
            }
        }
        List<String> ans = new ArrayList<String>();
        for(String s:common.keySet()){
            if(common.get(s) == min)
                ans.add(s);
        }
        String sans[] = new String[ans.size()];
        for(int i=0;i<sans.length;i++)
            sans[i] = ans.get(i);
        return sans;
    }
}
